internal/internal_nexus_task_handler.go

```diff
@@ -0,0 +1,353 @@
+		}
+		// Default to internal error.
+		return nil, h.internalError(err), nil
+	}
+	switch t := opres.(type) {
+	case *nexus.HandlerStartOperationResultAsync:
+		return &nexuspb.Response{
+			Variant: &nexuspb.Response_StartOperation{
+				StartOperation: &nexuspb.StartOperationResponse{
+					Variant: &nexuspb.StartOperationResponse_AsyncSuccess{
+						AsyncSuccess: &nexuspb.StartOperationResponse_Async{OperationId: t.OperationID},
+					},
+				},
+			},
+		}, nil, nil
+	default:
+		// *nexus.HandlerStartOperationResultSync is generic, we can't type switch unfortunately.
+		value := reflect.ValueOf(t).Elem().FieldByName("Value").Interface()
+		payload, err := h.dataConverter.ToPayload(value)
+		if err != nil {
+			return nil, h.internalError(fmt.Errorf("cannot convert nexus sync result: %w", err)), nil
+		}
+		return &nexuspb.Response{
+			Variant: &nexuspb.Response_StartOperation{
+				StartOperation: &nexuspb.StartOperationResponse{
+					Variant: &nexuspb.StartOperationResponse_SyncSuccess{
+						SyncSuccess: &nexuspb.StartOperationResponse_Sync{
+							Payload: payload,
+						},
+					},
+				},
+			},
+		}, nil, nil
+	}
+}
+
+func (h *nexusTaskHandler) handleCancelOperation(ctx context.Context, nctx *NexusOperationContext, req *nexuspb.CancelOperationRequest, header nexus.Header) (*nexuspb.Response, *nexuspb.HandlerError, error) {
+	cancelOptions := nexus.CancelOperationOptions{Header: header}
+	var err error
+	func() {
+		defer func() {
+			recovered := recover()
+			if recovered != nil {
+				var ok bool
+				err, ok = recovered.(error)
+				if !ok {
+					err = fmt.Errorf("panic: %v", recovered)
+				}
+
+				nctx.Log.Error("Panic captured while handling nexus task", tagStackTrace, string(debug.Stack()), tagError, err)
+			}
+		}()
+		err = h.nexusHandler.CancelOperation(ctx, req.GetService(), req.GetOperation(), req.GetOperationId(), cancelOptions)
+	}()
+	if ctx.Err() != nil {
+		return nil, nil, ctx.Err()
+	}
+	if err != nil {
+		var handlerErr *nexus.HandlerError
+		if errors.As(err, &handlerErr) {
+			return nil, nexusHandlerErrorToProto(handlerErr), nil
+		}
+		// Default to internal error.
+		return nil, h.internalError(err), nil
+	}
+
+	return &nexuspb.Response{
+		Variant: &nexuspb.Response_CancelOperation{
+			CancelOperation: &nexuspb.CancelOperationResponse{},
+		},
+	}, nil, nil
+}
+
+func (h *nexusTaskHandler) internalError(err error) *nexuspb.HandlerError {
+	h.logger.Error("error processing nexus task", "error", err)
+	return nexusHandlerError(nexus.HandlerErrorTypeInternal, "internal error")
+}
+
+func (h *nexusTaskHandler) goContextForTask(nctx *NexusOperationContext, header nexus.Header) (context.Context, context.CancelFunc, *nexuspb.HandlerError) {
+	// Associate the NexusOperationContext with the context.Context used to invoke operations.
+	ctx := context.WithValue(context.Background(), nexusOperationContextKey, nctx)
+
+	timeoutStr := header.Get(nexus.HeaderRequestTimeout)
```

Comment:
Hrmm, I might have expected server to handle timeout and send cancellation. If a handler chooses not to respect timeout, what happens? If it is also handled server side, I think it's best to not also do it here except maybe with some considerable leeway to ensure server's cancellation logic is the one always processed.

Reply:
We don't have sticky execution so there's not a way for the server to send cancelation. The server propagates this from the client request and also has its own context deadline.

I think it's good to cancel work that we know can't complete in time and have the server propagate this timeout.
As for whether the context deadline in the SDK should be shorter/same/longer than the one tracked on the server, fair point, but maybe shorter here is better so the SDK doesn't get a false sense of completion and the metrics we emit can be more accurate.

Reply:
But what happens in racy cases where SDK and server hit at the same time? If server-side happens first does that look the exact same as if the client-side one hit first and reported this failure back? It's important to have one system be the arbiter of true timeout errors. If you want a just-in-case for the other system, no prob, but having two separate systems that race each other to report timeout failure can result in racy inconsistencies.

Reply:
Keep in mind that clients also set deadlines on the request context.
I think giving the most up-to-date and accurate timeout to all of the processes involved in handling this request is preferable. That's how gRPC does it and this is essentially the SDK handling RPCs.

Also note that on context deadline errors we don't respond to the server, we just drop the task so I'm not as concerned with the racy inconsistency you're talking about.